File: frontend/src/services/authService.ts
import axios from "axios";

import { BASE_URL } from "@/data/constants";
import { axiosInstance } from "@/services/axiosInstance";
import { AuthRequest, AuthResponse } from "@/types/auth";

class AuthService {
  private baseUrl: string = "/api/v1/auth";

  public async register(user: AuthRequest) {
    const { data } = await axiosInstance.post<AuthResponse>(`${this.baseUrl}/register`, user);
    this.setTokens(data);
    return data;
  }

  public async login(user: AuthRequest) {
    const { data } = await axiosInstance.post<AuthResponse>(`${this.baseUrl}/login`, user);
    this.setTokens(data);
    return data;
  }

  public async refreshToken() {
    console.log("refreshToken");
    const token = localStorage.getItem("refreshToken");
    this.removeTokens();
    const { data } = await axios.post<AuthResponse>(
      `${BASE_URL}${this.baseUrl}/refresh-token`,
      null,
      {
        headers: {
          Authorization: token ? `Bearer ${token}` : undefined
        }
      }
    );
    this.setTokens(data);
    return data;
  }

  public async logout() {
    this.removeTokens();
  }

  private setTokens(data: AuthResponse) {
    localStorage.setItem("accessToken", data.accessToken);
    localStorage.setItem("refreshToken", data.refreshToken);
  }

  private removeTokens() {
    localStorage.removeItem("accessToken");
    localStorage.removeItem("refreshToken");
  }
}

export const authService = new AuthService();

File: backend/src/main/java/com/nellshark/backend/services/AuthService.java
package com.nellshark.backend.services;

import com.nellshark.backend.dtos.requests.AuthRequest;
import com.nellshark.backend.dtos.responses.AuthResponse;
import com.nellshark.backend.models.User;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.lang.NonNull;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@Slf4j
@RequiredArgsConstructor
public class AuthService {

  private final UserService userService;
  private final CaptchaService captchaService;
  private final JwtService jwtService;
  private final AuthenticationManager authenticationManager;
  private final PasswordEncoder passwordEncoder;

  public AuthResponse register(@NonNull AuthRequest authRequest) {
    final String userEmail = authRequest.email();
    log.info("Registering user: email='{}'", userEmail);

    userService.checkEmailAvailability(authRequest.email());

    captchaService.verifyRecaptcha("clientCaptchaToken");

    final String encodedPassword = passwordEncoder.encode(authRequest.password());
    User user = new User(userEmail, encodedPassword);
    userService.saveUser(user);

    String accessToken = jwtService.generateToken(user);
    String refreshToken = jwtService.generateRefreshToken(user);

    return new AuthResponse(accessToken, refreshToken);
  }

  public AuthResponse login(@NonNull AuthRequest request) {
    final String userEmail = request.email();
    log.info("Authenticating user: email='{}'", userEmail);

    authenticationManager.authenticate(
        new UsernamePasswordAuthenticationToken(
            userEmail,
            request.password()
        )
    );

    User user = userService.getUserByEmail(request.email());
    String accessToken = jwtService.generateToken(user);
    String refreshToken = jwtService.generateRefreshToken(user);

    return new AuthResponse(accessToken, refreshToken);
  }

  public AuthResponse refreshToken(@NonNull String authHeader) {
    log.info("Refreshing token");

    if (StringUtils.isBlank(authHeader) || !authHeader.startsWith("Bearer ")) {
      throw new BadCredentialsException("Invalid or missing Authorization header");
    }

    String refreshToken = authHeader.substring(7);
    String userEmail = jwtService.extractUsername(refreshToken);

    if (StringUtils.isBlank(userEmail)) {
      throw new BadCredentialsException("Unable to extract user email from refresh token");
    }

    User user = userService.getUserByEmail(userEmail);
    if (!jwtService.isTokenValid(refreshToken, user)) {
      throw new BadCredentialsException("Refresh token is invalid");
    }

    String newAccessToken = jwtService.generateToken(user);
    String newRefreshToken = jwtService.generateRefreshToken(user);

    return new AuthResponse(newAccessToken, newRefreshToken);
  }
}
